File: src/Core/Game.hs
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Redundant bracket" #-}
{-# HLINT ignore "Redundant return" #-}
{-# HLINT ignore "Use list literal pattern" #-}
{-# HLINT ignore "Use null" #-}
module Core.Game where

import Controllers.MatchesController
import Data.Time.Clock (getCurrentTime, NominalDiffTime, UTCTime, diffUTCTime)
import Interface.DrawBoard
import Utils.Validator
import Control.Concurrent (threadDelay)
import System.Console.ANSI
import System.IO
import Controllers.BoardController
import Utils.Utils as UT
import Controllers.AccountsController
import Controllers.PlayerController
import Data.Char
import Controllers.LettersController

{- Trata o input do usuário obtido durante o jogo
 Recebe: uma match
 Recebe: o arrau de todas as palavras do português
 Recebe: a string de input do usuário
 Retorna: a match atualizada de acordo com as ações realoizadas e uma string que informa o status da operação realizada-}
fluxHandler :: Match -> [String] -> String -> IO (Match, String)
-- CASOS ESPECIAIS : :C PAUSAR PARTIDA
fluxHandler match wl ":c" = fluxHandler match wl ":C"
fluxHandler match _ ":C" = return (match, "\n" ++ (map toUpper (accName (pAcc (getPlayerOnTurn match)))) ++ " pausou o jogo!\n")

-- CASOS ESPECIAIS: :! PULAR TURNO
fluxHandler match _ ":!" = do
                        return (skipPlayerTurn match, ">> " ++ (map toUpper (accName (pAcc (getPlayerOnTurn match)))) ++ " pulou o turno!\n") 

-- CASOS ESPECIAIS: :! VER MANUAL
fluxHandler match wordlist ":?" = do
                        UT._printRulesShortened
                        return (match, "")
-- CASOS ESPECIAIS: :* trocar letra : sem letra
fluxHandler match w (':':'*':[]) = do
                        return (match, "")
-- CASOS ESPECIAIS: :* trocar letra : letras válidas e inválidas
fluxHandler match w (':':'*':t) = 
                    -- obtém o objeto letter a partir de um char
                    let lttr = UT.getLetterObject (toUpper (head t))
                    in case lttr of
                    Just letter -> do --se é um caracter válido
                        let hasLttr = playerHasLetter match letter
                        if hasLttr then do -- se o jogador tem o caracter
                            switched <- switchPlayerLetter match letter
                            return (skipPlayerTurn switched, ((map toUpper (accName (pAcc (getPlayerOnTurn match)))) ++ " trocou uma letra.\n"))
                        else do
                            UT.__colorText (" > Escolha uma letra válida \n") Blue
                            c <- getLine
                            fluxHandler match w (":*" ++ c)
                    Nothing -> do -- o caracter é inválido
                        UT.__colorText (" > Escolha uma letra válida \n") Blue
                        c <- getLine
                        fluxHandler match w (":*" ++ c)

-- CASO COMUM: player tenta colocar uma palavra no tabuleiro                      
fluxHandler match wordlist input 
    -- coordenadas válidas, palavras existentes e letras válidas
    |(res && ((length palavrasInvalidas) == 0) && (length letrasInvalidas) == 0) = do
        updatedPlayer <- updatePlayerLetters (removePlayerLetters (incPlayerScore (resetMatchSkipsQtd (updateMatchBoard match boardAtualizado)) points) letrasUsadas)
        let m = toggleMatchTurn updatedPlayer
        return (m, ("Palavra válida! Pontos: " ++ (show points) ++ "\n"))
    -- coordenadas válidas mas letras inválidas (player não tem as letras)
    |(res && ((length letrasInvalidas) /= 0)) = do
        UT.__colorText ("\nVocê não tem as letras: " ++ (show letrasInvalidas) ++ " \nTente novamente: \n") Red
        putStr "Digite sua palavra no formato X00 V/H PALAVRA:\n > "
        hFlush stdout
        i <- getLine
        (m, msg)  <- (fluxHandler match wordlist i)
        return (m, msg)
    --coordenadas válidas mas forma palavras que não existem
    |(res && ((length palavrasInvalidas) /= 0)) = do
        UT.__colorText ("\nPalavras inválidas: " ++ (show palavrasInvalidas) ++ " \nTente novamente: \n") Red
        putStr "Digite sua palavra no formato X00 V/H PALAVRA:\n > "
        hFlush stdout
        i <- getLine
        (m, msg) <- (fluxHandler match wordlist i)
        return (m, msg)
    --coordenadas inválidas
    | otherwise = do
        UT.__colorText "\nCoordenada ou Formatação inválidas, tente novamente: \n" Red
        putStr "Digite sua palavra no formato X00 V/H PALAVRA:\n > "
        hFlush stdout
        i <- getLine
        (m, msg) <- (fluxHandler match wordlist i)
        return (m, msg)
    where 
        (res, palavrasInvalidas, points, letrasInvalidas, letrasUsadas) = (initialValidation match wordlist input)
        boardAtualizado = (updateBoard (placeWord (readWordInput input match) (mBoard match)))



{- Loop principal do jogo, correspondendo a um turno: recebe input do jogador, valida o input e informa resultados sobre a tentativa
 Recebe: uma match em um estado inicial
 Recebe: o array de todas as palavras da língua portuguesa
 Recebe: uma variável de controle de tempo
 Recebe: a string informando o estado da últimna execução do método (último turno) 
 Retorna: uma match atualizada de acordo com as ações realizadas no método-}
gameLoop :: Match -> [String] -> UTCTime -> String -> IO (Match)
gameLoop match wordList lastUpdate lastMessage = do
    -- Verifica uma das condições de parada:
    -- 4 saltos de turno/trocas de letra seguidos
    -- A partida ficar sem letra para dar aos jogadores
    if mSkips match == 4 || length (mLetters match) == 0 then do
        finishedMatch <- finishMatch match
        return finishedMatch
    else do
        -- Mostra a tela de transição e aguarda continuação
        clearScreen
        UT.__colorText lastMessage Green
        hFlush stdout
        UT.__colorText "> Enter para ver o tabuleiro do jogador da vez!\n\n" Blue
        hFlush stdout
        c <- getLine

        -- Mostra a tela de jogo 
        printBoard match
        UT.__colorText ("Turno de: " ++ (map toUpper (accName (pAcc (getPlayerOnTurn match))))) Blue
        putStr "\nDigite sua palavra no formato X00 V/H PALAVRA:\n > "
        hFlush stdout
        
        -- Recebe o input do jogador e valida
        input <- getLine

        -- Jogador pausou a partida e saiu para o menu
        if input == ":C" || input == ":c" then do 
            UT.__colorText "\n >> Pausando e saindo do jogo...\n" Green
            UT.__colorText " > Aperte enter...\n\n" Blue
            c <- getLine
            return match
        -- Jogador jogou uma palavra ou outra ação especial
        else do 
            currentTime <- getCurrentTime
            let elapsed = realToFrac (currentTime `diffUTCTime` lastUpdate) :: NominalDiffTime
            let updatedTimer = mTimer match - realToFrac elapsed

            -- Se tiver acabado o tempo não registra a palavra jogada
            if updatedTimer <= 0 then do
                let updatedMatch = toggleMatchTurn match
                updateMatchJson updatedMatch
                gameLoop updatedMatch wordList currentTime "\nTempo de rodada excedido!\n"
            
            -- Se estiver dentro do tempo recebe a palavra ou comando e os processa
            else do
                (m, msg) <- fluxHandler match wordList input
                if getPlayerOnTurn match /= getPlayerOnTurn m then do
                    updateMatchJson m
                    gameLoop m wordList currentTime msg
                else do 
                    let updatedMatch = updateMatchTimer m updatedTimer
                    threadDelay 100000
                    updateMatchJson updatedMatch
                    gameLoop updatedMatch wordList currentTime msg
            

File: prolog/Core/game.pl

game_loop(MatchName, LastMessage):-
    match_exists(MatchName),
    match(MatchName, _, _, _, _, MatchLetters, _, MatchTimer, MatchSkips),
    get_turn_player_name(MatchName, PlayerOnTurn),
    update_player_letters(MatchName),
    ((MatchSkips == 4 ; length(MatchLetters, 0)) -> 
        (   
            ansi_format([bold, fg(green)], '>> 4 skips ou trocas! Encerrando o jogo...\n\n',[]),
            ansi_format([bold, fg(blue)], 'Aperte Enter para ver o placar...\n\n',[]),
            no_period_input(_),
            finish_match(MatchName)
        );(

            % Mostra a tela de transição e aguarda continuação
            string_upper(PlayerOnTurn, PlayerOnTurnUpper),
            ansi_format([bold, fg(blue)], '~w' , [LastMessage]),
            ansi_format([bold, fg(green)], '> Enter para ver o tabuleiro de ~w!\n\n\n', [PlayerOnTurnUpper]),
            no_period_input(I),
            
            % Mostra a tela de jogo 
            print_board(MatchName),
            now(StartTime),
            writef('Digite sua palavra no formato X00 V/H PALAVRA:\n > '),

            % Recebe o input do jogador e valida
            no_period_input(UserPlayString),

            % Jogador pausou a partida e saiu para o menu
            ((UserPlayString == ':C'; UserPlayString == ':c') -> 
                (
                    ansi_format([bold, fg(green)], '\n >> Pausando e saindo do jogo...\n',[]),
                    ansi_format([bold, fg(blue)], ' > Aperte enter...\n\n',[]),
                    no_period_input(_),
                    !
                ); 
                
                % Jogador jogou uma palavra ou outra ação especial
                (   
                    now(CurTime),
                    (
                        (
                            too_long(StartTime, CurTime),
                            toggle_player_turn(MatchName),
                            game_loop(MatchName, '\nTempo de rodada excedido!\n'),!
                        ) ; (
                            flux_handler(MatchName, UserPlayString, Msg, StartTime),
                            game_loop(MatchName, Msg),!
                        )
                    )
                )
            )
        )
    ).


flux_handler(MatchName, ':!', Msg,_):-
    get_turn_player_name(MatchName, PlayerOnTurn),
    skip_player_turn(MatchName),
    format(atom(Msg), '>> ~w pulou o turno!\n', [PlayerOnTurn]).

flux_handler(MatchName, ':?', Msg,_):-
    print_short_rules,
    print_board(MatchName),
    writef('Digite sua palavra no formato X00 V/H PALAVRA:\n > '),
    no_period_input(I),
    flux_handler(MatchName, I, Msg).

flux_handler(_, ':*', '',_).

flux_handler(MatchName, S, Msg,_):- 
    sub_string(S, 0, 2, _, ':*'), 
    string_chars(S, SL),
    nth0(2, SL, L),
    get_turn_player_name(MatchName, P),
    get_player_letters(MatchName, P, PlayerLetters),
    letter_score(L, Score),

    ((
        Score > -1, member(L, PlayerLetters),
        format(atom(Msg), ' >> ~w trocou a letra ~w!\n\n', [P, L]),
        switch_player_letter(MatchName, L),
        skip_player_turn(MatchName)
    );(
        ansi_format([bold, fg(red)],' > Escolha uma letra válida \n', []),
        write(':*'),
        no_period_input(I),
        format(atom(A), ':*~w', [I]),
        flux_handler(MatchName, A, Msg))), !.


flux_handler(MatchName,StringInput, Msg, StartTime):-
    validation(MatchName, StringInput, [true, Points, UsedLetters, [], []]),
    now(CurTime),
    ((too_long(StartTime,CurTime),
    Msg = '\nTempo de rodada excedido!\n'),!;
    (remove_player_letters(MatchName, UsedLetters),
    inc_player_score(MatchName, Points),
    get_match_board_name(MatchName, BoardName),
    update_cur_tiles(BoardName),
    reset_match_skips(MatchName),
    format(atom(Msg), '\nPalavra válida! Pontos: ~d\n', [Points]))),
    toggle_player_turn(MatchName), !.


flux_handler(MatchName,StringInput, Msg, StartTime):-
    validation(MatchName, StringInput, [false|_]),
    now(CurTime),
    ((too_long(StartTime,CurTime),
        Msg = '\nTempo de rodada excedido!\n',
        toggle_player_turn(MatchName), !),!
        ;
    (get_match_board_name(MatchName, BoardName),
    reset_work_tiles(BoardName),
    ansi_format([bold, fg(red)], '\nJogada inválida, tente novamente: \n', []),
    write('Digite sua palavra no formato X00 V/H PALAVRA:\n > '),
    no_period_input(I),
    flux_handler(MatchName, I, Msg, StartTime))),!.

flux_handler(MatchName,StringInput, Msg):-
    validation(MatchName, StringInput, [_, _, _, Invalidletters, _]),
    get_match_board_name(MatchName, BoardName),
    reset_work_tiles(BoardName),
    ansi_format([bold, fg(red)], '\nVocê não tem as letras: ~w, tente novamente: \n', [InvalidLetters]),
    write('Digite sua palavra no formato X00 V/H PALAVRA:\n > '),
    no_period_input(I),
    flux_handler(MatchName, I, Msg), !.

flux_handler(MatchName,StringInput, Msg):-
    validation(MatchName, StringInput,[ _, _, _, _, InvalidWords]),
    get_match_board_name(MatchName, BoardName),
    reset_work_tiles(BoardName),
    ansi_format([bold, fg(red)], '\nPalavras inválidas: ~w, tente novamente: \n', [InvalidWords]),
    write('Digite sua palavra no formato X00 V/H PALAVRA:\n > '),
    no_period_input(I),
    flux_handler(MatchName, I, Msg), !.


flux_handler(_,_,'Pânico geral!').

